File: java/dsprotocol/src/riocomm2/RioCommProtocol.java
package riocomm2;

import java.net.*;

public class RioCommProtocol {

    InetAddress mRioAddress;
    IOHandler ioHandler;

    public RioCommProtocol(int teamnum, String fallbackAddress) throws UnknownHostException{
        mRioAddress = getAddress(teamnum, fallbackAddress);
        ioHandler = new IOHandler();
    }

    public void start() throws Exception{
        ioHandler.start(mRioAddress);
        if(ioHandler.getLastException() != null)
            throw ioHandler.getLastException();
    }

    public Exception getExceptions(){
        return ioHandler.getLastException();
    }

    public void close(){
        ioHandler.close();
    }

    public SendingPacket getSending(){
        return ioHandler.getSendingPacket();
    }

    public void setSending(SendingPacket sending){
        ioHandler.setSendingPacket(sending);
    }

    public void wantEnabled(boolean enabled){
        getSending().setWantEnabled(enabled);
    }

    public void wantFMS(boolean FMS){
        getSending().setWantFMS(FMS);
    }

    public void wantEstop(boolean estop){
        getSending().setWantEstop(estop);
    }

    public void wantReset(boolean reset){
        getSending().setWantReset(reset);
    }

    public void wantRestart(boolean restart){
        getSending().setWantRestart(restart);
    }

    public void setMode(SendingPacket.ControlMode mode){
        getSending().setMode(mode);
    }

    public SendingPacket.ControlMode getMode(){
        return getSending().getMode();
    }

    public RecievingPacket getRecieving(){
        return ioHandler.getRecievingPacket();
    }

    private InetAddress getAddress(int teamNum, String fallback) throws UnknownHostException {
        InetAddress rioAddress = null;
        try {
            rioAddress = InetAddress.getByName("roborio-" + teamNum + "-frc.local");
            System.out.println("Rio found at: " + rioAddress.getHostAddress());
            return rioAddress;
        } catch (UnknownHostException e) {
            System.out.println("System Error: could not find RIO with hostname '" + e.getMessage() + "'");
            try {
                rioAddress = InetAddress.getByName(fallback);
                System.out.println("Rio found at: " + rioAddress.getHostAddress());
                return rioAddress;
            } catch (UnknownHostException except1) {
                System.out.println("System Error: could not find RIO with address '" + e.getMessage() + "'");
            }
        }
        throw new UnknownHostException();
    }

    private class IOHandler{

        private static final int sendingPort = 1110;
        private static final int recievingPort = 1150;
        private InetAddress mToRio;
        private boolean wantStop = false;
        private Exception mLastException = null;
        private DatagramSocket sock;
        private Thread runner;
        private SendingPacket sending = SendingPacket.defaultPacket;
        private RecievingPacket recieving;
        private byte[] recieve;

        private final Runnable run = () -> {
            while(!isValidError() && !wantStop) {
                try {
                    sock = new DatagramSocket(recievingPort);
                    sock.setSoTimeout(10);
                    byte[] toSend;
                    while (!wantStop) {
                        //sending
                        toSend = sending.getPacket();
                        DatagramPacket toRio = new DatagramPacket(toSend, toSend.length, mToRio, sendingPort);
                        sock.send(toRio);

                        //wait for response
                        Thread.sleep(18);

                        //receive into system
                        DatagramPacket fromRio = new DatagramPacket(recieve, recieve.length);
                        sock.receive(fromRio);
                        recieving.setData(fromRio.getData());


                    }
                } catch (Exception e) {
                    mLastException = e;
                    sock.close();
                }
            }
        };


        public void start(InetAddress address){
            mToRio = address;
            recieve = new byte[64];
            recieving = new RecievingPacket(64);
            runner = new Thread(run);
            runner.start();
        }

        public void close(){
            wantStop = true;
            if(sock != null) sock.close();
        }

        public void setSendingPacket(SendingPacket packet){
            sending = packet;
        }

        public SendingPacket getSendingPacket(){
            return sending;
        }

        public RecievingPacket getRecievingPacket(){
            return recieving;
        }

        public Exception getLastException() {
            if(isValidError()) {
                Exception prep = mLastException;
                mLastException = null;
                return prep;
            }
            return null;
        }

        public boolean getStopped(){
            return wantStop;
        }

        private boolean isValidError(){
            return mLastException != null &&
                    !((mLastException instanceof InterruptedException) ||
                    (mLastException instanceof SocketTimeoutException));
        }

    }

}

File: Unity DS/Assets/scripts/RioCommProtocol.cs

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Threading;

namespace comm {
	public class RioCommProtocol {

		public static void startComm(int teamNum, String fallback) {
			Console.WriteLine("starting comm protocol");
			IPAddress resolved = resolveRio(teamNum, fallback);
			Console.WriteLine("starting IOhandler thread");
			if(resolved != null) IOHandler.start(resolved);
		}

		public static void close() {
			IOHandler.close();
		}

		public static SendingPacket getSending() {
			return IOHandler.getSendingPacket();
		}

		public static void setSending(SendingPacket sending) {
			IOHandler.setSendingPacket(sending);
		}

		public static void wantEnabled(bool enabled) {
			getSending().setWantEnabled(enabled);
		}

		public static void wantFMS(bool FMS) {
			getSending().setWantFMS(FMS);
		}

		public static void wantEstop(bool estop) {
			getSending().setWantEstop(estop);
		}

		public static void wantReset(bool reset) {
			getSending().setWantReset(reset);
		}

		public static void wantRestart(bool restart) {
			getSending().setWantRestart(restart);
		}

		public static void setMode(SendingPacket.ControlMode mode) {
			getSending().setMode(mode);
		}

		public static SendingPacket.ControlMode getMode() {
			return getSending().getMode();
		}

		public static RecievingPacket getRecieving() {
			return IOHandler.getRecievingPacket();
		}

		private static IPAddress resolveRio(int teamNum, String fallback) {
			IPAddress[] resolved;
			try {
				resolved = Dns.GetHostAddresses("roborio-" + teamNum + "-frc.local");
				if (resolved.Length > 0) {
					Console.WriteLine("Rio found at " + resolved[0]);
					return resolved[0];
				}
			}
			catch (SocketException ex) {
				Console.WriteLine("RIO not found with DNS");
			}

			try {
				resolved = Dns.GetHostAddresses(fallback);
				if (resolved.Length > 0) {
					Console.WriteLine("Rio found at " + resolved[0]);
					return resolved[0];
				}
			}
			catch (SocketException ex) {
				Console.WriteLine("RIO not found with fallback IP");
			}

			throw new Exception("Could not Resolve with DNS or fallback", new SocketException());
		}

		private class IOHandler {
			private const int sendingPort = 1110;
			private const int recievingPort = 1150;
			private static SendingPacket sending = new SendingPacket();
			private static RecievingPacket recieving = new RecievingPacket(64);
			private static byte[] recieve = new byte[64];
			private static UdpClient udpClient;
			private static IPAddress rioAddress;
			private static IPEndPoint RioWithPort;
			private static IPEndPoint recieveFromRio;
			private static bool wantStop = false;
			//private static List<SendingPacket.Tag> TCPEvents;

			private static void handleComm() {

				while (!wantStop) {
					try{
						udpClient = new UdpClient(recievingPort);
						//udpClient.Client.Blocking = false;
						byte[] toSend;
						while (!wantStop)
						{
							//sending
							toSend = sending.getPacket();
							//Console.WriteLine("sending");
							udpClient.Send(toSend, toSend.Length, RioWithPort);

							//wait for response
							Thread.Sleep(18);

							//receive into system
							//Console.WriteLine("recieving");
							if (udpClient.Available > 0)
							{
								recieve = udpClient.Receive(ref recieveFromRio);
								//Console.WriteLine("recieved");
								recieving.setData(recieve);
							}
							
						}
					}
					catch (Exception e)
					{
						Console.Out.WriteLine("Encountered an exception while operating com protocol: " + e);
					}
					finally
					{
						udpClient?.Close();
					}
				}

			}

			public static void start(IPAddress address) {
				rioAddress = address;
				RioWithPort = new IPEndPoint(rioAddress, sendingPort);
				recieveFromRio = new IPEndPoint(rioAddress, sendingPort);
				Thread commThread = new Thread(handleComm);
				commThread.Start();
			}

			public static void close() {
				wantStop = true;
				Console.WriteLine("aborting Rio comm");
			}

			public static void setSendingPacket(SendingPacket packet) {
				sending = packet;
			}

			public static SendingPacket getSendingPacket() {
				return sending;
			}

			public static RecievingPacket getRecievingPacket() {
				return recieving;
			}

		}
	}

}
